dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -419,6 +430,19 @@ private LambdaEvent createBasicEvent(String repository, String gitReference, Str
         return updatedServices;
     }
 
+    /**
+     * Decide whether a gitReference is excluded, given a workflow/service's filters
+     * @param gitRefPath Path.of(gitReference) for glob matching with PathMatcher
+     * @param filters Filters specified for a workflow/service in .dockstore.yml
+     * @return
+     */
+    private boolean filterGitReference(final Path gitRefPath, final List<String> filters) {
```

Comment:
1. I think this should go somewhere else -- if we ever add dockstore.yml support to tools, this code wouldn't be accessible.
2. I think this method is a good candidate for unit tests, where you could throw a bunch of patterns it at to make sure they work as expected.
3. CircleCI filters have separate subsections for branches and tags. We're not going with that syntax, which is fine, but I think we need to have that functionality expressable. Maybe it's already here and I just don't know the pattern of a gitRefPath, but for example, is it possible to:
    * Filter for all tags but no branches
    * Filter for all releases, no branches, and no tags that aren't also releases (this one is a stretch and may not be possible)
    * Other patterns that I can't think of?

If these are possible, could you give examples in comments and/or the unit tests? If not, I think we need to handle them, in particular the only filter tags case.

Reply:
> 1. I think this should go somewhere else -- if we ever add dockstore.yml support to tools, this code wouldn't be accessible.

I'll look into that

> 2. I think this method is a good candidate for unit tests, where you could throw a bunch of patterns it at to make sure they work as expected.

True, that would have been a good alternative to the integration test I made...

> 3. CircleCI filters have separate subsections for branches and tags. We're not going with that syntax, which is fine, but I think we need to have that functionality expressable. Maybe it's already here and I just don't know the pattern of a gitRefPath, but for example, is it possible to:

After discussion with Denis above & my most recent commit:

>    * Filter for all tags but no branches

`tags/*`
Other way around: `heads/*`

>    * Filter for all releases, no branches, and no tags that aren't also releases (this one is a stretch and may not be possible)

Not sure how to tell whether a git reference is a release, unless releases have consistent tag naming.
```
- tags/1.0.*
- tags/1.1.*
...
```

>    * Other patterns that I can't think of?

[PathMatcher docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String)) detail how to make more advanced globs.
I'm currently automatically prefixing filter globs with `refs/{,heads/,tags/}`.
Here is a handy site I found for experimenting with globbing patterns: https://globster.xyz/

Reply:
> True, that would have been a good alternative to the integration test I made...

I would argue it's a good complement to integration tests, not either/or. If you want to test 20 different patterns, it's going to be time consuming and require a lot of set up with integration tests (create a branch named x, create a branch named y, create a tag named z, etc.).